mesonbuild/compilers/compilers.py

```diff
@@ -805,6 +805,7 @@ def compile(self, code: 'mesonlib.FileOrString',
                                        'testfile.' + self.default_suffix)
                 with open(srcname, 'w', encoding='utf-8') as ofile:
                     ofile.write(code)
+                    ofile.write('\n')
```

Comment:
@dcbaker this will technically have a visible effect on user code in meson.build I think -- depending on whether user code passed to `cc.compiles()` is well formed. Do we care? I *think* we don't care. The alternative is indeed adding a newline to 20 different locations across 5 different files.

Reply:
To which I'm wondering, would it be safer to do something like:
```python
with open(...) as f:
    f.write(code)
    if not code.endswith('\n'):
        f.write('\n')
```

Reply:
Even if it _is_ harmless (from a POSIX and compiler perspective) to add an extra newline when there is already one there, what you suggest @dcbaker may be preferable anyway.  To me it makes it more obvious to the reader _why_  we might be appending a newline.  I'm happy to update this PR if that's the consensus.

Reply:
> I sure hope we don't care. But now that I think about it, is there a case where some language is going to get mad in the other way? like does (for example) D have a `--pedantic-errors` option that will yell if there's two trailing newlines?, ala:
> 
> ```meson
> has_int = dc.compiles('int add(int l, int r) { return l + r; }\n')
> ```
> 
> since we'd now insert a second newline?

D is pretty lax when it comes to warnings, it will only complain on stuff that it knows is 99-100% wrong from a running code standpoint. The only stylistic diagnostic that I'm aware of is using an empty statement with `for` or `if` and that's an error, not a warning:
```d
void main () {
    if (true);
	int oops;
}
```
```
a.d(2): Error: use `{ }` for an empty statement, not `;`
```

Reply:
Good to know about D :) I just I don't want to end up in a position where we start inserting an extra newline in the check case to fix warnings/errors on *nix, then start having user's code fail in other cases because their compiler/platform is very strict that there must be *exactly one* newline at the end of the file.

Reply:
@eli-schwartz, what do you think?